transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java

```diff
@@ -327,23 +346,54 @@ private void doWriteMultiple(ChannelOutboundBuffer in) {
      }
 
     protected final void doWriteSingle(ByteBuf buf) {
+        assert (ioState & WRITE_SCHEDULED) == 0;
         IOUringSubmissionQueue submissionQueue = submissionQueue();
         submissionQueue.addWrite(socket.intValue(), buf.memoryAddress(), buf.readerIndex(),
                 buf.writerIndex());
         ioState |= WRITE_SCHEDULED;
     }
 
     //POLLOUT
-    private void addPollOut() {
+    private void schedulePollOut() {
         assert (ioState & POLL_OUT_SCHEDULED) == 0;
-        ioState |= POLL_OUT_SCHEDULED;
         IOUringSubmissionQueue submissionQueue = submissionQueue();
         submissionQueue.addPollOut(socket.intValue());
+        ioState |= POLL_OUT_SCHEDULED;
+    }
+
+    void schedulePollRdHup() {
+        assert (ioState & POLL_RDHUP_SCHEDULED) == 0;
+        IOUringSubmissionQueue submissionQueue = submissionQueue();
+        submissionQueue.addPollRdHup(fd().intValue());
+        ioState |= POLL_RDHUP_SCHEDULED;
     }
 
     abstract class AbstractUringUnsafe extends AbstractUnsafe {
         private IOUringRecvByteAllocatorHandle allocHandle;
 
+        @Override
+        public void close(ChannelPromise promise) {
+            if ((ioState & (WRITE_SCHEDULED | READ_SCHEDULED | CONNECT_SCHEDULED)) == 0) {
+                forceClose(promise);
```

Comment:
An alternative to delay enqueueing the close operation, could be to enqueue it with the `IOSQE_IO_DRAIN` flag. Then the close operation will not be started until prior operations have completed, though following operations will also not be started until the close operation has completion, thus creating a slight queue hiccup. The hiccup might not be a big deal, though, compared to the state tracking we otherwise have to do.

Reply:
The problem with this is that this is not only about the "close sys call" but also about other state that is set in netty itself. So this will not help at all :)

Reply:
it causes an issue if we set IOSQE_IO_DRAIN, when some fd submit another fd polling event before, sqe wouldn't be executed as it waits until the socket is readable